tests/migrations/test_optimizer.py

```diff
@@ -270,6 +270,35 @@ def test_optimize_through_create(self):
             app_label="testapp",
         )
 
+        # This could be optimized a bit more but at least it generates a valid
+        # set of operations.
+        self.assertOptimizesTo(
+            [
+                migrations.CreateModel('Book', [('name', models.CharField(max_length=255))]),
+                migrations.CreateModel('Person', [('name', models.CharField(max_length=255))]),
+                migrations.AddField('book', 'author', models.ForeignKey('test_app.Person', models.CASCADE)),
+                migrations.CreateModel('Review', [('book', models.ForeignKey('test_app.Book', models.CASCADE))]),
+                migrations.CreateModel('Reviewer', [('name', models.CharField(max_length=255))]),
+                migrations.AddField('review', 'reviewer', models.ForeignKey('test_app.Reviewer', models.CASCADE)),
+                migrations.RemoveField('book', 'author'),
+                migrations.DeleteModel('Person'),
+            ],
+            [
+                migrations.CreateModel('Person', [('name', models.CharField(max_length=255))]),
+                migrations.CreateModel('Book', [
+                    ('name', models.CharField(max_length=255)),
+                    ('author', models.ForeignKey('test_app.Person', models.CASCADE)),
+                ]),
+                migrations.CreateModel('Reviewer', [('name', models.CharField(max_length=255))]),
+                migrations.CreateModel('Review', [
+                    ('book', models.ForeignKey('test_app.Book', models.CASCADE)),
+                    ('reviewer', models.ForeignKey('test_app.Reviewer', models.CASCADE)),
+                ]),
+                migrations.RemoveField('book', 'author'),
```

Comment:
The ideal optimization of this scenario would be

```python
[
    migrations.CreateModel('Book', [('name', models.CharField(max_length=255))]),
    migrations.CreateModel('Reviewer', [('name', models.CharField(max_length=255))]),
    migrations.CreateModel('Review', [
        ('book', models.ForeignKey('test_app.Book', models.CASCADE)),
        ('reviewer', models.ForeignKey('test_app.Reviewer', models.CASCADE)),
    ]),
]
```

But the optimizer blocks the `CreateModel('Book') & RemoveField('Book', 'author')` reduction because the in between `CreateModel('Review')` operation references `Book`. If this reduction was to happen then the `CreateModel('Person') & DeleteModel('Person')` reduction could be performed as the in between `CreateModel('Book')` operation wouldn't be referencing `Person` anymore.

@rtpg I think this is the perfect example of optimizations that requires two passes to be achieved.